core/src/main/java/org/stellar/anchor/sep38/Sep38Service.java

```diff
@@ -360,21 +360,39 @@ public Sep38QuoteResponse postQuote(JwtToken token, Sep38PostQuoteRequest reques
 
     // Calculate amounts: sellAmount = buyAmount * totalPrice
     BigDecimal bTotalPrice = decimal(rate.getTotalPrice());
-    BigDecimal bSellAmount, bBuyAmount;
-    if (request.getSellAmount() != null) {
-      bSellAmount = decimal(request.getSellAmount());
-      bBuyAmount = bSellAmount.divide(bTotalPrice, buyAsset.getDecimals(), RoundingMode.HALF_UP);
-    } else {
-      bBuyAmount = decimal(request.getBuyAmount());
+    BigDecimal bSellAmount =
+        request.getSellAmount() != null ? decimal(request.getSellAmount()) : null;
+    BigDecimal bBuyAmount = request.getBuyAmount() != null ? decimal(request.getBuyAmount()) : null;
+
+    // Use the rate's buy and sell amounts if they were returned
+    if (rate.getSellAmount() != null) {
+      bSellAmount = decimal(rate.getSellAmount());
+    }
+    if (rate.getBuyAmount() != null) {
+      bBuyAmount = decimal(rate.getBuyAmount());
+    }
+
+    // This should not happen because we previously checked at least one was not null
+    if (bBuyAmount == null && bSellAmount == null) {
+      throw new ServerErrorException(
+          String.format(
```

Comment:
Can you change the message that they are both null?

Reply:
Done